src/clj/puppetlabs/services/master/master_core.clj

```diff
@@ -873,6 +873,127 @@
          :headers {"Content-Type" "application/octet-stream"}
          :body (get-code-content environment code-id file-path)}))))
 
+(defn is-bolt-project?
+  [dir]
+  (or (fs/exists? (str dir "/bolt-project.yaml"))
+      (fs/exists? (str dir "/bolt.yaml"))
+      (fs/exists? (str dir "/Boltdir"))))
+
+(defn find-bolt-project
+  [project-dir project-name]
+  (->> project-dir
+       io/file
+       .list
+       seq
+       (filter #(is-bolt-project? (str project-dir "/" %)))
+       (some #{project-name})))
+
+(defn get-project-root
+  [project-dir project-name]
+  (let [boltdir-path (str project-dir "/" project-name "/Boltdir")]
+    (if (fs/exists? boltdir-path)
+      boltdir-path
+      (str project-dir "/" project-name))))
+
+(defn- dirs-in-project-modulepath
+  "List all directories in a bolt project's modulepath"
+  [project-root]
+  (->> ["modules" "site-modules" "site"]
+       (map #(str project-root "/" %))
+       (filter fs/exists?)
+       (mapcat fs/list-dir)
+       (filter fs/directory?)))
+
+(defn find-project-module
+  "Given the path of a project and the name of a module, search the project's
+  module path and return the path to that module, or nil."
+  [project-root module]
+  (->> project-root
+       dirs-in-project-modulepath
+       (filter #(= module (fs/base-name %)))
+       first))
+
+(defn mount->path-component
+  "Map the fileserving \"mount\" to the subdirectory of modules it will serve
+  files out of."
+  [mount]
+  (case mount
+    "modules" "files"
+    ;; default to itself
+    mount))
```

Comment:
I probably need to be explicit about which are allowed, i think right now you could probably request out of any subdirectory because of this

Reply:
+1

Reply:
Nevermind, this is already handled through the routing table and the error is returned in `handle-project-file-content`